code/modules/power/cell.dm

```diff
@@ -149,28 +149,33 @@
 					corrupt()
 
 /obj/item/stock_parts/cell/attack_self(mob/user)
-	var/mob/living/carbon/human/H = user
-	var/datum/species/ethereal/eth_species = H.dna?.species
-	
-	if(istype(H) && istype(eth_species))			
+	if(isethereal(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/species/ethereal/E = H.dna.species
+		if(E.drain_time > world.time)
+			return
 		if(charge < 100)
 			to_chat(H, "<span class='warning'>The [src] doesn't have enough power!</span>")
 			return
-		if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL - 5)
+		var/obj/item/organ/stomach/ethereal/stomach = H.getorganslot(ORGAN_SLOT_STOMACH)
+		if(!istype(stomach))
+			to_chat(H, "<span class='warning'>You can't receive charge!</span>")
+			return
+		if(stomach.crystal_charge >= ETHEREAL_CHARGE_FULL)
 			to_chat(H, "<span class='warning'>Your charge is full!</span>")
 			return
 		to_chat(H, "<span class='notice'>You clumsily channel power through the [src] and into your body, wasting some in the process.</span>")
-		if(do_after(user, 50, target = src))
-			if((charge >= 100) && (eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL - 5))
+		E.drain_time = world.time + 20
+		if(istype(stomach))

```

Comment:
You already checked if stomach was oftype on line 161, so just remove this check